contracts/kill_switch/KillSwitch.sol

```diff
@@ -0,0 +1,142 @@
+    }
+
+    IIssuesRegistry public defaultIssuesRegistry;
+    mapping (address => bool) internal allowedInstances;
+    mapping (address => bool) internal deniedBaseImplementations;
+    mapping (bytes32 => IssuesSettings) internal appsIssuesSettings;
+
+    event DefaultIssuesRegistrySet(address indexed issuesRegistry);
+    event AllowedInstanceSet(address indexed instance, bool allowed);
+    event DeniedBaseImplementationSet(address indexed base, bool denied);
+    event IssuesRegistrySet(bytes32 indexed appId, address issuesRegistry);
+    event HighestAllowedSeveritySet(bytes32 indexed appId, IIssuesRegistry.Severity severity);
+
+    function initialize(IIssuesRegistry _defaultIssuesRegistry) external onlyInit {
+        initialized();
+        _setDefaultIssuesRegistry(_defaultIssuesRegistry);
+    }
+
+    function setDefaultIssuesRegistry(IIssuesRegistry _defaultIssuesRegistry)
+        external
+        auth(SET_DEFAULT_ISSUES_REGISTRY_ROLE)
+    {
+        _setDefaultIssuesRegistry(_defaultIssuesRegistry);
+    }
+
+    function setAllowedInstance(address _instance, bool _allowed)
+        external
+        authP(SET_ALLOWED_INSTANCES_ROLE, arr(_instance))
+    {
+        allowedInstances[_instance] = _allowed;
+        emit AllowedInstanceSet(_instance, _allowed);
+    }
+
+    function setDeniedBaseImplementation(address _base, bool _denied)
+        external
+        authP(SET_DENIED_BASE_IMPLS_ROLE, arr(_base))
+    {
+        deniedBaseImplementations[_base] = _denied;
+        emit DeniedBaseImplementationSet(_base, _denied);
+    }
+
+    function setIssuesRegistry(bytes32 _appId, IIssuesRegistry _issuesRegistry)
+        external
+        authP(SET_ISSUES_REGISTRY_ROLE, arr(_appId))
+    {
+        require(isContract(_issuesRegistry), ERROR_ISSUES_REGISTRY_NOT_CONTRACT);
+        appsIssuesSettings[_appId].issuesRegistry = _issuesRegistry;
+        emit IssuesRegistrySet(_appId, address(_issuesRegistry));
+    }
+
+    function setHighestAllowedSeverity(bytes32 _appId, IIssuesRegistry.Severity _severity)
+        external
+        authP(SET_HIGHEST_ALLOWED_SEVERITY_ROLE, arr(_appId))
+    {
+        appsIssuesSettings[_appId].highestAllowedSeverity = _severity;
+        emit HighestAllowedSeveritySet(_appId, _severity);
+    }
+
+    /**
+     * @dev Note that we are not checking if the appId, base address and instance address are valid and if they correspond
+     *      to each other in order to reduce extra calls. However, since this is only a query method, wrong input
+     *      can only result in invalid output. Internally, this method is used from the Kernel to stop calls if needed,
+     *      and we have several tests to make sure its usage is working as expected.
+     */
+    function shouldDenyCallingApp(bytes32 _appId, address _base, address _instance) external returns (bool) {
+        // if the instance is allowed, then allow given call
+        if (isInstanceAllowed(_instance)) {
```

Comment:
Wonder if we should a check for `_instance == address(this)` here, so that all KillSwitch instances implicitly can't turn themselves off.

Reply:
Yea, I was wondering that too, but I finally decided to whitelist the kill switch instance from the factory instead. However, we could change this function as well, WDYT?

Reply:
I'm definitely thinking making it implicit here sounds like a better abstraction; a `KillSwitch` turning itself off could result in some weird nasty business (although again, right now nothing in `KillSwitch` is `killSwitchProtected` so nothing would happen, which is also quite awkward).